Translate: State as a snapshot page (#440)
reactjs/it.react.dev: src/content/learn/state-as-a-snapshot.md

---

> title: Lo stato come un'istantanea

Reviewer: ```suggestion
title: Lo State come un'Istantanea
```
Come da linee guida, non traduciamo alcuni termini come _state_

---

<Intro>

> Le variabili di stato possono sembrare delle normali variabili JavaScript su cui è possibile leggere e scrivere. Tuttavia lo stato si comporta più come un'istantanea. Quando lo si assegna, non si modifica la variabile di stato che si ha già, ma si innesca un nuovo renderizzato.

Reviewer: ```suggestion
Le variabili di state possono sembrare delle normali variabili JavaScript su cui è possibile leggere e scrivere. Tuttavia lo state si comporta più come un'istantanea. Quando lo si assegna, non si modifica la variabile di state che si ha già, ma si innesca una nuova renderizzazione.
```

</Intro>

<YouWillLearn>

> * Come l'assegnazione dello stato innesca re-renderizzazioni

Reviewer: ```suggestion
* Come l'assegnazione dello state innesca re-renderizzazioni
```

> * Come e quando viene aggiornato lo stato

Reviewer: ```suggestion
* Come e quando viene aggiornato lo state
```

> * Perché lo stato non viene aggiornato immediatamente dopo averlo assegnato

Reviewer: ```suggestion
* Perché lo state non viene aggiornato immediatamente dopo averlo assegnato
```

> * Come i gestori di eventi accedono ad un'"istantanea" dello stato

Reviewer: ```suggestion
* Come i gestori di eventi accedono ad un'"istantanea" dello state
```

</YouWillLearn>

> ## Come l'assegnazione dello stato innesca re-renderizzazioni {/*setting-state-triggers-renders*/}

Reviewer: ```suggestion
## Come l'assegnazione dello state innesca renderizzazioni {/*setting-state-triggers-renders*/}
```

> Si potrebbe pensare che l'interfaccia dell'utente cambi direttamente in risposta ad un evento dell'utente stesso, come un click. In React funziona in modo leggermente diverso da questo modello mentale. Nella pagina precedente si è visto che quando [si assegna lo stato si richiede un nuovo renderizzato](/learn/render-and-commit#step-1-trigger-a-render) da React. Questo significa che per far reagire l'interfaccia al evento, è neccessario *aggiornare lo stato*.

Reviewer: ```suggestion
Si potrebbe pensare che l'interfaccia dell'utente cambi direttamente in risposta ad un evento dell'utente stesso, come un click. In React funziona in modo leggermente diverso da questo modello mentale. Nella pagina precedente si è visto che quando [si assegna lo state si richiede una nuova re-renderizzazione](/learn/render-and-commit#step-1-trigger-a-render) da React. Questo significa che per far reagire l'interfaccia al evento, è necessario *aggiornare lo state*.
```

> In questo esempio, quando premi "Send", `setIsSent(true)` dice a React di re-renderizzare la UI.

Reviewer: ```suggestion
In questo esempio, quando premi "send", `setIsSent(true)` dice a React di re-renderizzare la UI:
```

</Sandpack>

Ecco cosa succede quando fai click sul pulsante:

1. Viene eseguito il gestore di eventi `onSubmit`.

> 2. `setIsSent(true)` assegnat `isSent` a `true` e mette in coda un nuovo render.

Reviewer: ```suggestion
2. `setIsSent(true)` assegna `isSent` a `true` e mette in coda un nuovo render.
```

3. React re-renderizza il componente in base al nuovo valore di `isSent`.

> Esaminiamo più da vicino la relazione tra lo stato e il renderizzato.

Reviewer: ```suggestion
Esaminiamo più da vicino la relazione tra lo state e il renderizzato.
```

> ## La renderizzazione scatta un'istantanea nel tempo {/*la-renderizzazione-scatta-unistantanea-nel-tempo*/}

Reviewer: ```suggestion
## La renderizzazione scatta un'istantanea nel tempo {/*rendering-takes-a-snapshot-in-time*/}
```
Attenzione: i link non vanno tradotti

> ["Renderizzare"](/learn/render-and-commit#step-2-react-renders-your-components) significa che React chiama il componente, che è una funzione. Il JSX che restituisce tale funzione è una istantanea della UI nel tempo. Le props, i gestori di eventi e le variabili locali sono stati calcolati **utilizzando il suo stato al momento del renderizzato.**

Reviewer: ```suggestion
["Renderizzare"](/learn/render-and-commit#step-2-react-renders-your-components) significa che React chiama il componente, che è una funzione. Il JSX che restituisce tale funzione è come un'istantanea della UI nel tempo. Le props, i gestori di eventi e le variabili locali sono stati calcolati **utilizzando il suo state al momento della renderizzazione.**
```

> A differenza di una fotografia o di un fotogramma di un film, l'"istantanea" della UI che viene restituita è interattiva. Include la logica, come i gestori di eventi che specificano cosa succede in risposta agli input. React aggiorna lo schermo in base a questa istantanea e collega i gestori di eventi. Di conseguenza, la pressione di un pulsante attiverà il gestore di click dal JSX.

Reviewer: ```suggestion
A differenza di una fotografia o di un fotogramma di un film, l'"istantanea" della UI che viene restituita è interattiva. Include la logica, come i gestori di eventi che specificano cosa succede in risposta agli input. React aggiorna lo schermo in base a questa istantanea e collega i gestori di eventi. Di conseguenza, la pressione di un pulsante attiverà il gestore di click dal tuo JSX.
```

Quando React re-renderizza un componente:

1. React chiama di nuovo la tua funzione.
2. La tua funzione restituisce una nuova istantanea JSX.
3. React aggiorna quindi la schermata in modo che corrisponda all'istantanea restituita.

<IllustrationBlock sequential>
    <Illustration caption="React esegue la funzione" src="/images/docs/illustrations/i_render1.png" />

>     <Illustration caption="Calcola la istantanea" src="/images/docs/illustrations/i_render2.png" />

Reviewer: ```suggestion
    <Illustration caption="Calcola l'istantanea" src="/images/docs/illustrations/i_render2.png" />
```

    <Illustration caption="Aggiorna l'albero del DOM" src="/images/docs/illustrations/i_render3.png" />
</IllustrationBlock>

> Come memoria di un componente, lo stato non è come una normale variabile che scompare dopo che la tua funzione restituisce un valore. Lo stato "vive" nello stesso React--come se si trattasse di uno scaffale!--fuori dalla tua funzione. Quando React chiama il tuo componente, fornisce un'istantanea della UI per quel particolare renderizzato. Il tuo componente restituisce un'istantanea della UI con un nuovo set di props e gestori di eventi nel suo JSX, tutti calcolati **usando i valori dello stato di quel renderizzato.**

Reviewer: ```suggestion
Come memoria di un componente, lo state non è come una normale variabile che scompare dopo che la tua funzione restituisce un valore. Lo state "vive" nello stesso React--come se si trattasse di uno scaffale!--fuori dalla tua funzione. Quando React chiama il tuo componente, fornisce un'istantanea della UI per quel particolare renderizzato. Il tuo componente restituisce un'istantanea della UI con un nuovo set di props e gestori di eventi nel suo JSX, tutti calcolati **usando i valori dello stato di quel renderizzato.**
```

<IllustrationBlock sequential>
  <Illustration caption="Tu dici a React di aggiornare lo state" src="/images/docs/illustrations/i_state-snapshot1.png" />

>   <Illustration caption="React aggiorna il valore dello stato" src="/images/docs/illustrations/i_state-snapshot2.png" />

Reviewer: ```suggestion
  <Illustration caption="React aggiorna il valore dello state" src="/images/docs/illustrations/i_state-snapshot2.png" />
```

>   <Illustration caption="React passa un'istantanea del valore dello stato al componente" src="/images/docs/illustrations/i_state-snapshot3.png" />

Reviewer: ```suggestion
  <Illustration caption="React passa un'istantanea del valore dello state al componente" src="/images/docs/illustrations/i_state-snapshot3.png" />
```

</IllustrationBlock>

> Ecco qui un piccolo esperimento per mostrarti come questo funziona. In questo esempio, potresti aspettarti che, cliccando il bottone "+3", il contatore venga incrementato tre volte, perché viene richiamato `setNumber(number + 1)` tre volte.

Reviewer: ```suggestion
Ecco qui un piccolo esperimento per mostrarti come questo funziona. In questo esempio, potresti aspettarti che, cliccando il bottone "+3", il contatore venga incrementato tre volte, perché viene chiamato `setNumber(number + 1)` tre volte.
```

</Sandpack>

Osserva che `number` viene incrementato solo una volta per click!

> **L'assegnazione dello stato cambia solo per il *prossimo* renderizzato.** Durante il primo renderizzato, `number` era `0`. È per questo che, nel gestore `onClick` di *quel renderizzato*, il valore di `number` è ancora `0`  anche dopo che stato richiamato `setNumber(number + 1)`:

Reviewer: ```suggestion
**L'assegnazione dello state cambia solo per la *prossima* renderizzazione .** Durante la prima renderizzazione, `number` era `0`. È per questo che, nel gestore `onClick` di *quella renderizzazione*, il valore di `number` è ancora `0`  anche dopo che stato chiamato `setNumber(number + 1)`:
```

}}>+3</button>
```

Questo è ciò che il gestore del click per questo pulsante dice a React di fare:

1. `setNumber(number + 1)`: `number` è `0` quindi `setNumber(0 + 1)`.
    - React si prepara per cambiare `number` a `1` nella prossima renderizzazione.
2. `setNumber(number + 1)`: `number` è `0` quindi `setNumber(0 + 1)`.
    - React si prepara per cambiare `number` a `1` nella prossima renderizzazione.
3. `setNumber(number + 1)`: `number` è `0` quindi `setNumber(0 + 1)`.
    - React si prepara per cambiare `number` a `1` nella prossima renderizzazione.

> Anche se chiami `setNumber(number + 1)` tre volte, nel gestore di eventi di *questa renderizzazione* `number` è sempre `0`, quindi stai assegnando lo stato a `1` per tre volte. Questo è il motivo per cui, dopo che il tuo gestore di eventi ha finito, React re-renderizza il componente con `number` uguale a `1` piuttosto che `3`.

Reviewer: ```suggestion
Anche se hai chiamato `setNumber(number + 1)` tre volte, nel gestore di eventi di *questa renderizzazione* `number` è sempre `0`, quindi stai assegnando lo state a `1` per tre volte. Questo è il motivo per cui, dopo che il tuo gestore di eventi ha finito, React ri-renderizza il componente con `number` uguale a `1` piuttosto che `3`.
```

> Puoi anche visualizzarlo sostituendo mentalmente le variabili di stato con i loro valori nel tuo codice. Poiché il valore della variabile di stato `number` è `0` per *questo renderizzato*, il suo gestore di eventi si presenta in questo modo:

Reviewer: ```suggestion
Puoi anche visualizzarlo sostituendo mentalmente le variabili di state con i loro valori nel tuo codice. Poiché il valore della variabile di state `number` è `0` per *questa renderizzazione*, il suo gestore di eventi si presenta in questo modo:
```

}}>+3</button>
```

> Per il prossimo renderizzato, `number` sarà `1`, quindi il gestore del click di *quel prossimo renderizzato* avrà l'aspetto seguente:

Reviewer: ```suggestion
Per la prossima renderizzazione, `number` sarà `1`, quindi il gestore del click di *quella prossima renderizzazione* avrà il seguente aspetto:
```

}}>+3</button>
```

Per questo, cliccando di nuovo il bottone, il contatore viene impostato a `2`, successivamente nel prossimo click a `3` e cosi via.

> ## Lo stato nel tempo {/*state-over-time*/}

Reviewer: ```suggestion
## Lo state nel tempo {/*state-over-time*/}
```

</Sandpack>

> Se utilizzi  il metodo si sostituzione di prima, puoi intuire che l'alert mostra "0":

Reviewer: ```suggestion
Se utilizzi  il metodo di sostituzione di prima, puoi intuire che l'alert mostra "0":
```

```js
setNumber(0 + 5);
alert(0);
```

> Ma cosa succede se imposti un timer per l'alert, in modo che si attivi _dopo_ che il componente ha re-renderizzato? Mostrerà "0" o "5"? Indovina!

Reviewer: ```suggestion
Ma cosa succede se imposti un timer per l'alert, in modo che si attivi _dopo_ che il componente viene ri-renderizzato? Mostrerà "0" o "5"? Indovina!
```

</Sandpack>

> Sorpreso? Se hai utilizzato il metodo di sostituzione, puoi vedere l'"istantanea" del valore dello stato passato all'alert.

Reviewer: ```suggestion
Sorpreso? Se hai utilizzato il metodo di sostituzione, puoi vedere l'"istantanea" del valore dello state passato all'alert.
```

}, 3000);
```

> Il valore dello stato memorizzato in React potrebbe essere cambiato al momento in cui si esegue l'alert, ma questo è stato pianificato utilizzando un'istantanea dello stato nel momento in cui l'utente ha interagito con il esso!

Reviewer: ```suggestion
Il valore dello state memorizzato in React potrebbe essere cambiato al momento in cui si esegue l'alert, ma questo è stato pianificato utilizzando un'istantanea dello state nel momento in cui l'utente ha interagito con esso!
```

> **Il valore di una variabile di stato non cambia mai all'interno di un renderizzamento,** anche se il codice del suo gestore di eventi è asincrono. Dentro l'`onClick` di *quel render*, il valore di `number` continua a essere `0` anche dopo che `setNumber(number + 5)` è stato eseguito. Il suo valore è stato "fissato" quando React ha "scattato l'istantanea" della UI chiamanto il tuo componente. 

Reviewer: ```suggestion
**Il valore di una variabile di state non cambia mai all'interno di una renderizzazione,** anche se il codice del suo gestore di eventi è asincrono. Dentro l'`onClick` di *quella renderizzazione*, il valore di `number` continua a essere `0` anche dopo che `setNumber(number + 5)` è stato eseguito. Il suo valore è stato "fissato" quando React ha "scattato l'istantanea" della UI chiamando il tuo componente. 
```

Ecco un esempio di come questo rende i gestori di eventi meno inclini a errori di sincronizzazione. Di seguito è riportato un formulario che invia un messaggio con un ritardo di cinque secondi. Immagina questo scenario:

1. Premi il pulsante "Send", inviando "Hello" ad Alice.
2. Prima dello scadere dei cinque secondi, cambia il valore del campo "To" in "Bob".

> Cosa ti aspetti che mostri l'alert? Mostrerà "You said Hello to Alice"? Oppure "You said Hello to Bob"? Fai una previsione basandoti su che ciò che hai imparato e dopo provalo:

Reviewer: ```suggestion
Cosa ti aspetti che mostri l'`alert`? Mostrerà "You said Hello to Alice"? Oppure "You said Hello to Bob"? Fai una previsione basandoti su che ciò che hai imparato e dopo provalo:
```

</Sandpack>

> **React mantiene i valori di stato "fissi" all'interno dei gestori di eventi di un renderizzato.** Non c'è bisogno di preoccuparsi se lo stato è cambiato durante l'esecuzione del codice.

Reviewer: ```suggestion
**React mantiene i valori di state "fissi" all'interno dei gestori di eventi di una renderizzazione.** Non c'è bisogno di preoccuparsi se lo state è cambiato durante l'esecuzione del codice.
```

> Ma cosa succede se vuoi leggere lo stato più recente prima di una nuova renderizzazione? In questo caso, si dovrebbe utilizzare una [funzione di aggiornamento dello stato](/learn/queueing-a-series-of-state-updates), descritta pagina successiva!

Reviewer: ```suggestion
Ma cosa succede se vuoi leggere lo state più recente prima di una nuova renderizzazione? In questo caso, si dovrebbe utilizzare una [funzione di aggiornamento dello state](/learn/queueing-a-series-of-state-updates), descritta pagina successiva!
```

<Recap>

> * L'assegnazione dello stato richiede un nuovo renderizzamento.

Reviewer: ```suggestion
* L'assegnazione dello state richiede una nuova renderizzazione.
```

> * React memorizza lo stato al di fuori del tuo componente, come se fosse su uno scaffale.

Reviewer: ```suggestion
* React memorizza lo state al di fuori del tuo componente, come se fosse su uno scaffale.
```

> * Quando chiami `useState`, React ti fornisce un'istantanea dello stato per quella renderizzazione.

Reviewer: ```suggestion
* Quando chiami `useState`, React ti fornisce un'istantanea dello state per quella renderizzazione.
```

> * Le variabili e i gestori di eventi non "sopravvivono" ai re-renderizzamenti. Ongi renderizzamento ha i propri gestori di eventi.

Reviewer: ```suggestion
* Le variabili e i gestori di eventi non "sopravvivono" alle ri-renderizzazioni. Ogni renderizzazione ha i propri gestori di eventi.
```

> * Ogni renderizzamento (e le funzioni al suo interno) "vedranno" sempre l'istantanea dello stato che React ha dato in *quel* renderizzamento.

Reviewer: ```suggestion
* Ogni renderizzazione (e le funzioni al suo interno) "vedrà" sempre l'istantanea dello state che React ha dato in *quella* renderizzazione.
```

> * Puoi sostituire mentalmente lo stato nei gestori di eventi, in modo simile a a come pensi nel JSX renderizzato. 

Reviewer: ```suggestion
* Puoi sostituire mentalmente lo state nei gestori di eventi, in modo simile a a come pensi nel JSX renderizzato. 
```

> * I gestori di eventi creati in passato hanno il valore di stato del renderizzato in cui sono stati creati

Reviewer: ```suggestion
* I gestori di eventi creati in passato hanno il valore di state della renderizzazione in cui sono stati creati
```

</Sandpack>

> Non fa differenza se lo imposti prima o dopo della chiamata `setWalk`. Il valore della renderizzazione di `walk` rimane fisso. La chiamata a `setWalk` lo modificherà solo per la prossima renderizzazione, ma non influenzerà il gestore di eventi di eventi della renderizzazione precendente.

Reviewer: ```suggestion
Non fa differenza se lo imposti prima o dopo della chiamata `setWalk`. Il valore della renderizzazione di `walk` rimane fisso. La chiamata a `setWalk` lo modificherà solo per la prossima renderizzazione, ma non influenzerà il gestore di eventi della renderizzazione precedente.
```

Questa riga potrebbe sembrare controintuitiva all'inizio:

```js
alert(walk ? 'Stop is next' : 'Walk is next');
```

> Però ha senso se la leggi come: "Se il semaforo mostra 'Walk now', il messaggio dovrebbe dire 'Stop is next.'" La variabile `walk` dentro il tuo gestore di eventi corrisponde al valore  di walk della renderizzazione e non cambia.

Reviewer: ```suggestion
Però ha senso se la leggi come: "Se il semaforo mostra 'Walk now', il messaggio dovrebbe dire 'Stop is next.'" La variabile `walk` dentro il tuo gestore di eventi corrisponde al valore  di `walk` della renderizzazione e non cambia.
```

</h1>
```

> Quindi, cliccando su "Change to Stop", si mette in coda un rendering con `walk` impostato su `false` e viene visualizzato "Stop is next". 

Reviewer: ```suggestion
Quindi, cliccando su "Change to Stop", si mette in coda una renderizzazione con `walk` impostato su `false` e viene visualizzato "Stop is next". 
```
